### ISSUE TITLE
Support for regex in `EffectFactory.matchPath`

### ISSUE THREAD

**Issue by zeucxb** [marblejs/marble#67]
### Can I use regex in the marble routes?

I know that are some especial chars that I can use like:

``` javascript
EffectFactory
  .matchPath('/:dir*')
```

But what about more complex regex?

``` javascript
EffectFactory
  .matchPath(/^\/(api|rest)\/.+$/)
```

**Comment by krzysztof-miemiec** [comment]
Internally Marble uses https://github.com/pillarjs/path-to-regexp which supports passing regexp to parameters. However everything has to be written as string. Have you tried `.matchPath('/(api|rest)/(.*)')`?

**Comment by krzysztof-miemiec** [comment]
If you're thinking about something more fancy, like writing actual regular expressions, it's not supported yet, but it's definitely doable. We'd have to think about regex pattern merging for nested routing. It's simple to get regex patterns and concatenate them, but I bet that sooner or later we'll get something like: `/my/parent/path/` and `/^\/(to-be|not-to-be)\/.+$/` and for any given route it will be simply evaluated as `false`, because of `^` in the middle. Then we'd have to remove one of the duplicated `\/` inside and so on... So, for now I came up with:
- get both parent and child path parts and simply concatenate them as string
- remove all `^` and `$` from that stringified regex (also using regex, what a perversion 😜 )
- `.replace(([\\/]{2,})/g, '/')`
- somehow extract all capturing groups and assign them to parameters; What should we do with groups that were defined in regex? There is a specification for named capturing groups that we could use, but AFAIK they haven't landed in JS. We can simply mark all groups from regex as non-capturing, but this has to be done earlier, before concatenation.
- append `^` and `$` once again
- probably check for possibility of hacks on such routing mechanism

Since Marble is one of the milion solutions available, the above is probably redundant and someone might have implemented such merger.

**Comment by JozefFlakus** [comment]
@zeucxb have you tested it with stringified Regex? 👇
https://github.com/marblejs/marble/issues/67#issuecomment-423290284

**Comment by zeucxb** [comment]
I tried and it does not work. Using as string it understand as a simple string and not a regex.

So,

`[200] http://localhost:4000/(api|rest)/(.*)`
`[404] http://localhost:4000/rest`
`[404] http://localhost:4000/rest/test`

**Comment by zeucxb** [comment]
Sorry, I have tried in the version that was running in the project here.

But checking the [urlParams.factory.ts](https://github.com/marblejs/marble/blob/master/packages/core/src/router/urlParams.factory.ts) it should work. So I created a new project with the current stable version and it works.

👎 Sorry 👎

**Comment by zeucxb** [comment]
But we have a problem if try to use regex and especial params.

ex.
```
matchPath: /(api|rest)/:file

request: /rest/test.json

params: { file: "rest" }
```

**Comment by zeucxb** [comment]
And without the special param _(only with regex)_ we cannot get url params.

**Comment by JozefFlakus** [comment]
re: https://github.com/marblejs/marble/issues/67#issuecomment-423666208
Can you post also the example regex / path that was used within `matchPath`?

**Comment by zeucxb** [comment]
@JozefFlakus  this?

**With** `:file`
```
matchPath: /(api|rest)/:file

request: /rest/test.json

params: { file: "rest" }
```

**Without** `:file`
```
matchPath: /(api|rest)/(.*)

request: /rest/test.json

params: { }
```

**Comment by JozefFlakus** [comment]
Thanks. Custom regex is problematic because params are mapped over declared names in order they occured, eg. `/(api|rest)/:file` has a one named parameter, which in result will be mapped to the first capturing grup. Maybe we should mach in more precisely @krzysztof-miemiec? 🤔

**Comment by krzysztof-miemiec** [comment]
@zeucxb I think you can try with a named param there, i.e.: `matchPath: /:type(api|rest)/:file`, then you should get `{type: "api", file: "test.json"}` as a result for `/api/test.json`.

The naming is probably a bug on our side, as **path-to-regexp** docs state that:
> It is possible to write an unnamed parameter that only consists of a matching group. It works the same as a named parameter, except it will be numerically indexed.

**Comment by zeucxb** [comment]
Yeah @krzysztof-miemiec it works. **Awesome!!!**

**Comment by JozefFlakus** [comment]
So far this kind of advanced routing is not documented and wasn't tested before. We should cover these features with test cases first. Thanks @zeucxb for pointing this out.

**Comment by zeucxb** [comment]
I was about to say that we should have routes' documentation. 😄

**Comment by zeucxb** [comment]
I was going to open another issue but I don't know if it's an expected behavior.

I've this code in my application:

```javascript
const web$ = combineRoutes("/admin", [
  EffectFactory
    .matchPath('/:dir')
    .matchType('GET')
    .use((req$) => ...), 
  EffectFactory
    .matchPath('*')
    .matchType('GET')
    .use((req$) => ... ),
]);
```

In this case a request to `/admin` gets 404 but `/admin/` works fine.

If I change the code to:

```javascript
const web$ = combineRoutes("/admin", [
  EffectFactory
    .matchPath('/:dir')
    .matchType('GET')
    .use((req$) => ...), 
  EffectFactory
    .matchPath('/')
    .matchType('GET')
    .use((req$) => ... ),
]);
```

Both `/admin` and `/admin/` works.

**Comment by JozefFlakus** [comment]
It is a bug. @zeucxb please create a separate issue so we can track it properly :)

**Comment by JozefFlakus** [comment]
@zeucxb fixed with [v1.1.1](https://github.com/marblejs/marble/releases/tag/v1.1.1) patch 💪
